### DISCUSSION TITLE
[15->16] psycopg2.errors.InvalidDatetimeFormat: invalid input syntax for type timestamp: ""

### DISCUSSION THREAD

**Discussion by trvlprogrammer** [OCA/OpenUpgrade#3812] [Q&A] [answered]
I have an error when migrating the db from 15 to 16. In this process, I am using openupgrade 16.0 

odoo command run : 
python odoo-bin -c ou_15_16.conf --database=buds_live_migrate --upgrade-path=D:\work\freelance\openupgrade\upenupgrade16\openupgrade_scripts\scripts --load=base,web,openupgrade_framework --update all --stop-after-init

ou_15_16.conf 
[options]
db_host = localhost
db_port = 5432
db_user = odoo15
db_password = odoo15
addons_path = D:\work\freelance\odoo16\odoo\odoo\addons,D:\work\freelance\odoo16\odoo\addons,D:\work\freelance\odoo16\odoo\freelance\kaika\app\root\mnt\enterprise-addons,D:\work\freelance\odoo16\odoo\freelance\kaika\app\src,D:\work\freelance\odoo16\odoo\upenupgrade16\openupgrade_framework,D:\work\freelance\odoo16\odoo\upenupgrade16\openupgrade_scripts

and here is the error that I got when running that command 

2023-04-05 02:57:33,202 12680 INFO buds_live_migrate odoo.modules.loading: Module iap loaded in 0.10s, 136 queries (+136 other)
2023-04-05 02:57:33,203 12680 INFO buds_live_migrate odoo.modules.loading: Loading module mail (24/148)
2023-04-05 02:57:33,278 12680 INFO buds_live_migrate odoo.modules.registry: module mail: creating or updating database tables
2023-04-05 02:57:44,790 12680 WARNING buds_live_migrate odoo.modules.loading: Transient module states were reset
2023-04-05 02:57:44,792 12680 ERROR buds_live_migrate odoo.modules.registry: Failed to load registry
2023-04-05 02:57:44,793 12680 CRITICAL buds_live_migrate odoo.service.server: Failed to initialize database `buds_live_migrate`.
Traceback (most recent call last):
  File "D:\work\freelance\odoo16\odoo\odoo\service\server.py", line 1299, in preload_registries
    registry = Registry.new(dbname, update_module=update_module)
  File "<decorator-gen-14>", line 2, in new
  File "D:\work\freelance\odoo16\odoo\odoo\tools\func.py", line 87, in locked
    return func(inst, *args, **kwargs)
  File "D:\work\freelance\odoo16\odoo\odoo\modules\registry.py", line 90, in new
    odoo.modules.load_modules(registry, force_demo, status, update_module)
  File "D:\work\freelance\odoo16\odoo\odoo\modules\loading.py", line 483, in load_modules
    processed_modules += load_marked_modules(cr, graph,
  File "D:\work\freelance\odoo16\odoo\odoo\modules\loading.py", line 371, in load_marked_modules
    loaded, processed = load_module_graph(
  File "D:\work\freelance\odoo16\odoo\odoo\modules\loading.py", line 207, in load_module_graph
    registry.init_models(cr, model_names, {'module': package.name}, new_install)
  File "D:\work\freelance\odoo16\odoo\odoo\modules\registry.py", line 521, in init_models
    model._auto_init()
  File "D:\work\freelance\odoo16\odoo\odoo\models.py", line 2574, in _auto_init
    new = field.update_db(self, columns)
  File "D:\work\freelance\odoo16\odoo\odoo\fields.py", line 985, in update_db
    self.update_db_column(model, column)
  File "D:\work\freelance\odoo16\odoo\odoo\fields.py", line 1021, in update_db_column
    self._convert_db_column(model, column)
  File "D:\work\freelance\odoo16\odoo\odoo\fields.py", line 1025, in _convert_db_column
    sql.convert_column(model._cr, model._table, self.name, self.column_type[1])
  File "D:\work\freelance\odoo16\odoo\odoo\tools\sql.py", line 128, in convert_column
    _convert_column(cr, tablename, columnname, columntype, using)
  File "D:\work\freelance\odoo16\odoo\odoo\tools\sql.py", line 147, in _convert_column
    cr.execute(query, log_exceptions=False)
  File "D:\work\freelance\odoo16\odoo\odoo\sql_db.py", line 313, in execute
    res = self._obj.execute(query, params)
psycopg2.errors.InvalidDatetimeFormat: invalid input syntax for type timestamp: ""

2023-04-05 02:57:44,796 12680 INFO buds_live_migrate odoo.service.server: Initiating shutdown
2023-04-05 02:57:44,796 12680 INFO buds_live_migrate odoo.service.server: Hit CTRL-C again or send a second signal to force the shutdown.
2023-04-05 02:57:44,797 12680 INFO buds_live_migrate odoo.sql_db: ConnectionPool(used=0/count=0/max=64): Closed 1 connections

Any solution for this? I am new at the migration odoo database. Any advice really appreciates. Thanks.

Regards,
Fatih

**Comment by pedrobaeza** [comment, marked as answer]
There aren't still scripts for migrating `mail` module.

**Reply by trvlprogrammer** [reply]
Ohh ok, thanks for the answer.

**Reply by aniol-roure** [reply]
Acording to the documentation on Module coverage 15.0 -> 16.0 (https://github.com/OCA/OpenUpgrade/blob/16.0/docsource/modules150-160.rst) Mail is now done, but I keep getting the same error as @trvlprogrammer. And I even checked redownloading the OpenUpgrade just in case some file went missing.
I'm not sure if I should be opening another post for that, but since it was the same as here that is why I'm asking. Any idea about what could be going wrong?

**Reply by pedrobaeza** [reply]
Yes, but that's a generic error coming from incorrect DB data. You have to identify in which column and record is happening, and see which data is incorrect there.

**Reply by aniol-roure** [reply]
Even tho I have not customized anything from this module? When I upgraded from Odoo 14 to 15 using OpenUpgrade I did not need to check for modules I hadn't customized. What I was thinking was that odoo default instalation might be using EN (US) thus dates are mm/dd/yy, meanwhile we are using dd/mm/yy. Is there a way to know what table, column and record gives the error without checking one by one? BTW, thanks for your help and fast answer!

**Reply by pedrobaeza** [reply]
Put your exact log here for reference.

**Reply by pedrobaeza** [reply]
Yeah, unless you put any "print" on the code to know which column is being converted, you don't have a clue what's going on. The error is that executing a query (which is also unknown on the piece of log), there's some content that is incorrect for a timestamp. It seems to be an empty value...

**Reply by aniol-roure** [reply]
Found it! Thank you very much for your help!
Here is the documentation that explains the changes in mail.mail: https://github.com/OCA/OpenUpgrade/pull/3832
As stated by @marielejeune: _Field scheduled_date on mail.mail is now datetime (not char anymore)_
I'll fix that and hopefully everything will go on smothly.
Again, @pedrobaeza thanks for your time.

**Reply by pedrobaeza** [reply]
Well, a possible fix may be to put as NULL values with "" (empty string) for avoiding that error and manual interventions.

**Reply by pedrobaeza** [reply]
Wait, that's already done on https://github.com/OCA/OpenUpgrade/blob/2e9d96f1288596f17d5af3fad84b86eca6dc3821/openupgrade_scripts/scripts/mail/16.0.1.10/pre-migration.py#L207 so it should work without doing anything more.

**Reply by aniol-roure** [reply]
Ì did:
UPDATE mail_mail SET scheduled_date = NULL WHERE scheduled_date = '';
and it worked (well it got stopped further away, but that is some other problem I'll have to investigate).
I dont know why the mail pre-migration did not work tho.

**Reply by pedrobaeza** [reply]
You should check, as maybe you are not executing cleanly the scripts.

**Reply by aniol-roure** [reply]
I'm running it using python 3.8 (checked requirentment.txt) with this config:
--addons-path=/home/vboxuser/Documents/odoo16/odoo/addons,/home/vboxuser/Documents/odoo16/odoo/custom-addons
--db-filter=test_db$
-c
/home/vboxuser/Documents/odoo16/odoo_local_nobom.conf
--update=all

and I droped openupgrade inside /home/vboxuser/Documents/odoo16/ so the scripts folder looks like this for example:
/home/vboxuser/Documents/odoo16/OpenUpgrade-16.0/openupgrade_scripts

Am I missing something?

**Reply by pedrobaeza** [reply]
I can't say, as there are a lot of variables. Check if there are traces in the log that migration scripts are being run.

**Reply by aniol-roure** [reply]
Yes, I think you where right and I wasn't full loading OpenUpgrade... My bad!
I was missing the --upgrade-path declaration as well as the path to the OpenUpgrade added in the --addons-path.
@pedrobaeza, thank you so much, that really helped!

**Comment by yboujraf** [comment]
Dear @aniol-roure 

Do you have the steps to fix this error message?

I just tried to upgrade from 15.0 to 16.0 and have same errors

Logs eroor : psycopg2.errors.InvalidDatetimeFormat: invalid input syntax for type timestamp: ""


> 2025-03-04T22:43:56.120755+01:00 odoo16 Odoo Server 16.0:myodoo.com:ERROR:odoo.modules.registry:Failed to load registry#012Traceback (most recent call last):https://github.com/OCA/OpenUpgrade/pull/12 File "/opt/odoo/16.0/odoo/odoo/modules/registry.py", line 87, in new#012 odoo.modules.load_modules(registry, force_demo, status, update_module)https://github.com/OCA/OpenUpgrade/pull/12 File "/opt/odoo/16.0/odoo/odoo/modules/loading.py", line 485, in load_modules#012 processed_modules += load_marked_modules(cr, graph,https://github.com/OCA/OpenUpgrade/pull/12 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^https://github.com/OCA/OpenUpgrade/pull/12 File "/opt/odoo/16.0/odoo/odoo/modules/loading.py", line 373, in load_marked_modules#012 loaded, processed = load_module_graph(https://github.com/OCA/OpenUpgrade/pull/12 ^^^^^^^^^^^^^^^^^^https://github.com/OCA/OpenUpgrade/pull/12 File "/opt/odoo/16.0/odoo/odoo/modules/loading.py", line 209, in load_module_graph#012 registry.init_models(cr, model_names, {'module': package.name}, new_install)https://github.com/OCA/OpenUpgrade/pull/12 File "/opt/odoo/16.0/odoo/odoo/modules/registry.py", line 526, in init_models#012 model._auto_init()https://github.com/OCA/OpenUpgrade/pull/12 File "/opt/odoo/16.0/odoo/odoo/models.py", line 2614, in _auto_init#012 new = field.update_db(self, columns)https://github.com/OCA/OpenUpgrade/pull/12 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^https://github.com/OCA/OpenUpgrade/pull/12 File "/opt/odoo/16.0/odoo/odoo/fields.py", line 1007, in update_db#012 self.update_db_column(model, column)https://github.com/OCA/OpenUpgrade/pull/12 File "/opt/odoo/16.0/odoo/odoo/fields.py", line 1043, in update_db_column#012 self._convert_db_column(model, column)https://github.com/OCA/OpenUpgrade/pull/12 File "/opt/odoo/16.0/odoo/odoo/fields.py", line 1047, in _convert_db_column#012 sql.convert_column(model._cr, model._table, self.name, self.column_type[1])https://github.com/OCA/OpenUpgrade/pull/12 File "/opt/odoo/16.0/odoo/odoo/tools/sql.py", line 128, in convert_column#012 _convert_column(cr, tablename, columnname, columntype, using)https://github.com/OCA/OpenUpgrade/pull/12 File "/opt/odoo/16.0/odoo/odoo/tools/sql.py", line 147, in _convert_column#012 cr.execute(query, log_exceptions=False)https://github.com/OCA/OpenUpgrade/pull/12 File "/opt/odoo/16.0/odoo/odoo/sql_db.py", line 321, in execute#012 res = self._obj.execute(query, params)https://github.com/OCA/OpenUpgrade/pull/12 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^#012psycopg2.errors.InvalidDatetimeFormat: invalid input syntax for type timestamp: ""


> 2025-03-04T22:43:56.122686+01:00 odoo16 Odoo Server 16.0:myodoo.com:CRITICAL:odoo.service.server:Failed to initialize database myodoo.com.#012Traceback (most recent call last):https://github.com/OCA/OpenUpgrade/pull/12 File "/opt/odoo/16.0/odoo/odoo/service/server.py", line 1323, in preload_registries#012 registry = Registry.new(dbname, update_module=update_module)https://github.com/OCA/OpenUpgrade/pull/12 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^https://github.com/OCA/OpenUpgrade/pull/12 File "", line 2, in new#012 File "/opt/odoo/16.0/odoo/odoo/tools/func.py", line 87, in locked#012 return func(inst, *args, **kwargs)https://github.com/OCA/OpenUpgrade/pull/12 ^^^^^^^^^^^^^^^^^^^^^^^^^^^https://github.com/OCA/OpenUpgrade/pull/12 File "/opt/odoo/16.0/odoo/odoo/modules/registry.py", line 87, in new#012 odoo.modules.load_modules(registry, force_demo, status, update_module)https://github.com/OCA/OpenUpgrade/pull/12 File "/opt/odoo/16.0/odoo/odoo/modules/loading.py", line 485, in load_modules#012 processed_modules += load_marked_modules(cr, graph,https://github.com/OCA/OpenUpgrade/pull/12 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^https://github.com/OCA/OpenUpgrade/pull/12 File "/opt/odoo/16.0/odoo/odoo/modules/loading.py", line 373, in load_marked_modules#012 loaded, processed = load_module_graph(https://github.com/OCA/OpenUpgrade/pull/12 ^^^^^^^^^^^^^^^^^^https://github.com/OCA/OpenUpgrade/pull/12 File "/opt/odoo/16.0/odoo/odoo/modules/loading.py", line 209, in load_module_graph#012 registry.init_models(cr, model_names, {'module': package.name}, new_install)https://github.com/OCA/OpenUpgrade/pull/12 File "/opt/odoo/16.0/odoo/odoo/modules/registry.py", line 526, in init_models#012 model._auto_init()https://github.com/OCA/OpenUpgrade/pull/12 File "/opt/odoo/16.0/odoo/odoo/models.py", line 2614, in _auto_init#012 new = field.update_db(self, columns)https://github.com/OCA/OpenUpgrade/pull/12 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^https://github.com/OCA/OpenUpgrade/pull/12 File "/opt/odoo/16.0/odoo/odoo/fields.py", line 1007, in update_db#012 self.update_db_column(model, column)https://github.com/OCA/OpenUpgrade/pull/12 File "/opt/odoo/16.0/odoo/odoo/fields.py", line 1043, in update_db_column#012 self._convert_db_column(model, column)https://github.com/OCA/OpenUpgrade/pull/12 File "/opt/odoo/16.0/odoo/odoo/fields.py", line 1047, in _convert_db_column#012 sql.convert_column(model._cr, model._table, self.name, self.column_type[1])https://github.com/OCA/OpenUpgrade/pull/12 File "/opt/odoo/16.0/odoo/odoo/tools/sql.py", line 128, in convert_column#012 _convert_column(cr, tablename, columnname, columntype, using)https://github.com/OCA/OpenUpgrade/pull/12 File "/opt/odoo/16.0/odoo/odoo/tools/sql.py", line 147, in _convert_column#012 cr.execute(query, log_exceptions=False)https://github.com/OCA/OpenUpgrade/pull/12 File "/opt/odoo/16.0/odoo/odoo/sql_db.py", line 321, in execute#012 res = self._obj.execute(query, params)https://github.com/OCA/OpenUpgrade/pull/12 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^#012psycopg2.errors.InvalidDatetimeFormat: invalid input syntax for type timestamp: ""

> 2025-03-04T22:43:56.122995+01:00 odoo16 Odoo Server 16.0:myodoo.com:INFO:odoo.service.server:Stopping gracefully

Thanks if you could help me

Best Regards,

**Reply by aniol-roure** [reply]
Yes, I had problems setting all the arguments needed for it to run as expected, and I had to try different orders apart from adding the missing ones. Make sure you have openupgradelib installed and your paramaters look like this:
--addons-path=_/path/to/addons,/path/to/OpenUpgrade_
--database=_db_name_
-c _/path/to/odoo.conf_
--log-level=debug
--upgrade-path=_/path/to/openupgrade_scripts/scripts_
--load=base,web,openupgrade_framework
--update all
--stop-after-init

After making sure this was working I got the real migration errors, and I had to solve conflicts with the db data, but the migration was running ok. 
Hope that helps.
Good luck.
